Fix panic caused by corner case in mempool

viquezclaudio opened the pull request:
- When adopted blocks are recieved and mempool update is called, we need to handle the case
  where the transactions in the adopted blocks caused some sender to be pruned from the
  accounts tree, because in this particular case, we need to remove all the txns that
  belong to the pruned sender from the mempool
- Some unittests were added to reproduce the issue and verify the fix
- This fixes #678

viquezclaudio committed:
Fix panic caused by corner case in mempool

- When adopted blocks are recieved and mempool update is called, we need to handle the case
  where the transactions in the adopted blocks caused some sender to be pruned from the
  accounts tree, because in this particular case, we need to remove all the txns that
  belong to the pruned sender from the mempool
- Some unittests were added to reproduce the issue and verify the fix
- This fixes #678

sisou commented on mempool/src/mempool.rs:
```diff
@@ -277,6 +277,9 @@ impl Mempool {
                                         "There is no account for this sender in the blockchain {}",
                                         tx.sender.to_user_friendly_address()
                                     );
+                                    // The account from this sender was pruned/removed, so we need to delete all acounts that belong to this sender
```
```suggestion
                                    // The account from this sender was pruned/removed, so we need to delete all transactions sent from this address
```

sisou commented on mempool/src/mempool.rs:
```diff
@@ -690,6 +693,21 @@ impl MempoolState {
         Some(tx)
     }
+
+    // Removes all the transactions that belong to some specific sender
```
This does not remove transactions _to_ the account, which I would say also _belong_ to it.
```suggestion
    // Removes all the transactions sent by some specific address
```

sisou commented on mempool/src/mempool.rs:
```diff
@@ -690,6 +693,21 @@ impl MempoolState {
+        }
+
+        //Finally remove all the state associated with this sender
+        self.state_by_sender.remove(&sender_address);
```
Isn't the state already removed above in line 700 when creating the loop range? Plus, there's a clippy warning on this line.

jsdanielh commented on mempool/tests/mod.rs:
```diff
@@ -912,6 +914,318 @@ async fn mempool_update() {
+
+    // Generate recipient accounts
+    let recipient_accounts = generate_accounts(recipient_balances, &mut genesis_builder, false);
+    // Use the same senders as before, because we want to test sender doesnt have enough funds to paid all pending txns
```
```suggestion
    // Use same senders as before, because we want to test that the sender doesn't have enough funds to pay all pending txns
```

jsdanielh commented on mempool/src/mempool.rs:
```diff
@@ -690,6 +692,18 @@ impl MempoolState {
         Some(tx)
     }
+
+    //Removes all the transactions sent by some specific address
```
```suggestion
    // Removes all the transactions sent by some specific address
```

jsdanielh commented on mempool/tests/mod.rs:
```diff
@@ -912,6 +914,318 @@ async fn mempool_update() {
+        .unwrap(),
+    ));
+
+    // Create mempool and subscribe with a custom txn stream.
```
```suggestion
    // Create mempool and subscribe with a custom txn stream
```

jsdanielh commented on mempool/tests/mod.rs:
```diff
@@ -912,6 +914,318 @@ async fn mempool_update() {
+        "Number of txns in the mempools is not what is expected"
+    );
+
+    //We need a block producer to produce blocks
```
```suggestion
    // We need a block producer to produce blocks
```

jsdanielh commented on mempool/tests/mod.rs:
```diff
@@ -912,6 +914,318 @@ async fn mempool_update() {
+
+    // Expect only 20 transations because in the adopted blocks we included 10 txns that would cause the senders
+    // to not have enough balance to pay for all txns already in the mempool.
+
```
Spare blank line

jsdanielh commented on mempool/tests/mod.rs:
```diff
@@ -912,6 +914,318 @@ async fn mempool_update() {
+    let updated_txns = mempool.get_transactions_for_block(10_000);
+
+    // Expect only 20 transations because in the adopted blocks we included 10 txns that would cause the senders
+    // to not have enough balance to pay for all txns already in the mempool.
```
```suggestion
    // to not have enough balance to pay for all txns already in the mempool
```

jsdanielh commented on mempool/tests/mod.rs:
```diff
@@ -912,6 +914,318 @@ async fn mempool_update() {
+            vec![0x41],
+        );
+
+        //We should succeed producing a block with the remaining mempool transactions
```
```suggestion
        // We should succeed producing a block with the remaining mempool transactions
```

jsdanielh commented on mempool/tests/mod.rs:
```diff
@@ -912,6 +914,318 @@ async fn mempool_update() {
+
+    // Generate recipient accounts
+    let recipient_accounts = generate_accounts(recipient_balances, &mut genesis_builder, false);
+    // Use the same senders as before, because we want to test sender doesnt have enough funds to paid all pending txns
```
```suggestion
    // Use same senders as before, because we want to test that the sender doesn't have enough funds to pay all pending txns
```

jsdanielh commented on mempool/tests/mod.rs:
```diff
@@ -912,6 +914,318 @@ async fn mempool_update() {
+        "Number of txns in the mempools is not what is expected"
+    );
+
+    //We need a block producer to produce blocks
```
```suggestion
    // We need a block producer to produce blocks
```

jsdanielh commented on mempool/tests/mod.rs:
```diff
@@ -912,6 +914,318 @@ async fn mempool_update() {
+            vec![0x41],
+        );
+
+        //We should succeed producing a block with the remaining mempool transactions
```
```suggestion
        // We should succeed producing a block with the remaining mempool transactions
```

viquezclaudio commented:
@jsdanielh changes applied

jsdanielh commented:
@viquezclaudio Also don't forget to sign the commits

viquezclaudio committed:
Include staking functionality for remove sender txns

- We also need to consider staking transactions when removing txns from some specific sender in the mempool
